File: test_samples/comprehensive/complex.go
// Comprehensive Go test file with complex structures
package complex

import (
	"fmt"
	"sync"
)

// Authenticator interface
type Authenticator interface {
	Authenticate(username, password string) bool
	Logout()
}

// User struct with embedding
type User struct {
	Username string
	ID       int
	mu       sync.RWMutex
}

// AdminUser embeds User
type AdminUser struct {
	User
	Permissions []string
}

// NewAdminUser constructor
func NewAdminUser(username string, id int) *AdminUser {
	return &AdminUser{
		User: User{
			Username: username,
			ID:       id,
		},
		Permissions: make([]string, 0),
	}
}

// Authenticate implements Authenticator
func (a *AdminUser) Authenticate(username, password string) bool {
	a.mu.RLock()
	defer a.mu.RUnlock()
	
	if username == "" || password == "" {
		return false
	}
	return a.Username == username && len(password) >= 8
}

// Logout implements Authenticator
func (a *AdminUser) Logout() {
	fmt.Println("Admin logged out")
}

// AddPermission adds a permission
func (a *AdminUser) AddPermission(perm string) {
	a.mu.Lock()
	defer a.mu.Unlock()
	a.Permissions = append(a.Permissions, perm)
}

// Generic session manager
type SessionManager[T any] struct {
	sessions map[string]T
	mu       sync.RWMutex
}

// NewSessionManager creates a new session manager
func NewSessionManager[T any]() *SessionManager[T] {
	return &SessionManager[T]{
		sessions: make(map[string]T),
	}
}

// CreateSession creates a new session
func (sm *SessionManager[T]) CreateSession(sessionID string, user T) {
	sm.mu.Lock()
	defer sm.mu.Unlock()
	sm.sessions[sessionID] = user
}

// GetSession retrieves a session
func (sm *SessionManager[T]) GetSession(sessionID string) (T, bool) {
	sm.mu.RLock()
	defer sm.mu.RUnlock()
	user, exists := sm.sessions[sessionID]
	return user, exists
}

// StatusCode type
type StatusCode int

const (
	Success StatusCode = iota
	Unauthorized
	Forbidden
	NotFound
)

// String method for StatusCode
func (sc StatusCode) String() string {
	switch sc {
	case Success:
		return "Success"
	case Unauthorized:
		return "Unauthorized"
	case Forbidden:
		return "Forbidden"
	case NotFound:
		return "NotFound"
	default:
		return "Unknown"
	}
}

File: test_samples/comprehensive/complex.rs
// Comprehensive Rust test file with complex structures
use std::collections::HashMap;
use std::sync::{Arc, RwLock};

/// Authenticator trait
pub trait Authenticator {
    fn authenticate(&self, username: &str, password: &str) -> bool;
    fn logout(&self);
}

/// Base User struct
#[derive(Debug, Clone)]
pub struct User {
    pub username: String,
    pub id: u32,
}

impl User {
    pub fn new(username: String, id: u32) -> Self {
        User { username, id }
    }
}

/// AdminUser with permissions
#[derive(Debug, Clone)]
pub struct AdminUser {
    user: User,
    permissions: Vec<String>,
}

impl AdminUser {
    pub fn new(username: String, id: u32) -> Self {
        AdminUser {
            user: User::new(username, id),
            permissions: Vec::new(),
        }
    }
    
    pub fn add_permission(&mut self, permission: String) {
        self.permissions.push(permission);
    }
    
    fn validate_password(&self, password: &str) -> bool {
        password.len() >= 8
    }
}

impl Authenticator for AdminUser {
    fn authenticate(&self, username: &str, password: &str) -> bool {
        if username.is_empty() || password.is_empty() {
            return false;
        }
        self.user.username == username && self.validate_password(password)
    }
    
    fn logout(&self) {
        println!("Admin logged out");
    }
}

/// Generic SessionManager
pub struct SessionManager<T> {
    sessions: Arc<RwLock<HashMap<String, T>>>,
}

impl<T: Clone> SessionManager<T> {
    pub fn new() -> Self {
        SessionManager {
            sessions: Arc::new(RwLock::new(HashMap::new())),
        }
    }
    
    pub fn create_session(&self, session_id: String, user: T) {
        let mut sessions = self.sessions.write().unwrap();
        sessions.insert(session_id, user);
    }
    
    pub fn get_session(&self, session_id: &str) -> Option<T> {
        let sessions = self.sessions.read().unwrap();
        sessions.get(session_id).cloned()
    }
}

/// Status code enum
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum StatusCode {
    Success = 200,
    Unauthorized = 401,
    Forbidden = 403,
    NotFound = 404,
}

impl StatusCode {
    pub fn message(&self) -> &'static str {
        match self {
            StatusCode::Success => "OK",
            StatusCode::Unauthorized => "Unauthorized",
            StatusCode::Forbidden => "Forbidden",
            StatusCode::NotFound => "Not Found",
        }
    }
}

/// Result type alias
pub type AuthResult<T> = Result<T, StatusCode>;

/// Complex function with error handling
pub fn authenticate_and_create_session<T: Authenticator + Clone>(
    user: &T,
    username: &str,
    password: &str,
    manager: &SessionManager<T>,
) -> AuthResult<String> {
    if user.authenticate(username, password) {
        let session_id = format!("session_{}", uuid::Uuid::new_v4());
        manager.create_session(session_id.clone(), user.clone());
        Ok(session_id)
    } else {
        Err(StatusCode::Unauthorized)
    }
}
